### PR TITLE
Use Actions to build and publish Docker images to Github Container Registry

### PR DESCRIPTION
**Pull request by ren-makes-things**
See #1003

### CODE REVIEW

**Comment by sudokoko** [comment]
I've approved check runs, if these are successful I will review after work.

**Comment by ren-makes-things** [comment]
also on pull requests IIRC before a merge the Docker / build workflow does not push to the github container registry, just builds to make sure it works, after merge the workflow reruns and pushes to container registry

**Comment by ren-makes-things** [comment]
if this workflow is merged to main

you should be able to run something like this `docker pull ghcr.io/LBPUnion/projectlighthouse:main` and get the docker image for the latest commit

File: .github/workflows/docker-publish.yml
Comment (sudokoko) [inline .github/workflows/docker-publish.yml:10]: What is the purpose of this?

File: .github/workflows/docker-publish.yml
Comment (sudokoko) [inline .github/workflows/docker-publish.yml:16]: Realistically, we should not be running this for PRs, opens the possibility for malware to be added without needing a merge.

**Review by sudokoko — CHANGES_REQUESTED** [review]
prelim review

File: .github/workflows/docker-publish.yml
Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:16]: It runs on PR's to test docker image building, but it does not push, see this comment someone over at gh made 

`
      # Build and push Docker image with Buildx (don't push on PR)
      # https://github.com/docker/build-push-action
`

File: .github/workflows/docker-publish.yml
Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:10]: to be honest im not sure, iirc this is meant to be a catch all template, so i'll just remove this because this is not something that needs to run scheduled

Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml]: removed

File: .github/workflows/docker-publish.yml
Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:14]: see this as well
```
        uses: docker/build-push-action@0565240e2d4ab88bba5387d719585280857ece09 # v5.0.0
        with:
          context: .
          push: ${{ github.event_name != 'pull_request' }}
``` 

signature tooling also gets skipped on PRs
```
      # Install the cosign tool except on PR
      # https://github.com/sigstore/cosign-installer
      - name: Install cosign
        if: github.event_name != 'pull_request'
        uses: sigstore/cosign-installer@e1523de7571e31dbe865fd2e80c5c7c23ae71eb4 #v3.4.0
        with:
          cosign-release: 'v2.2.3'

```

```
      # Sign the resulting Docker image digest except on PRs.
      # This will only write to the public Rekor transparency log when the Docker
      # repository is public to avoid leaking data.  If you would like to publish
      # transparency data even for private images, pass --force to cosign below.
      # https://github.com/sigstore/cosign
      - name: Sign the published Docker image
        if: ${{ github.event_name != 'pull_request' }}
        env:
          # https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-an-intermediate-environment-variable
          TAGS: ${{ steps.meta.outputs.tags }}
          DIGEST: ${{ steps.build-and-push.outputs.digest }}
        # This step uses the identity token to provision an ephemeral certificate
        # against the sigstore community Fulcio instance.
        run: echo "${TAGS}" | xargs -I {} cosign sign --yes {}@${DIGEST}
```

Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:14]: also when you ran the docker build workflow on this PR, it didn't push an image to the container registry, as https://github.com/orgs/LBPUnion/packages?repo_name=ProjectLighthouse is empty

Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:14]: Is this ok or are there any changes to this you want? @sudokoko

Comment (sudokoko) [reply .github/workflows/docker-publish.yml:14]: Let me check with the team to see if this is something that we'd be fine with including, I'll get back to you on this after I do a more thorough review

Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:14]: If there are any edits that need to be made let me know

Comment (sudokoko) [reply .github/workflows/docker-publish.yml:14]: I'm just going to push this aside for a bit, me personally I don't see any problems with this as long as it doesn't publish on PR.

File: .github/workflows/docker-publish.yml
Comment (sudokoko) [inline .github/workflows/docker-publish.yml:12]:
```suggestion
```

We don't necessarily use release tagging, this can be safely removed.

**Review by sudokoko — CHANGES_REQUESTED** [review]
Second wave of nitpicks, otherwise looks mostly good.

**Comment by ren-makes-things** [comment]
Done @sudokoko, ignore that one commit message being messed up, phone keyboard did a stupid

**Review by sudokoko — APPROVED** [review]

File: .github/workflows/docker-publish.yml
Comment (Zaprit) [inline .github/workflows/docker-publish.yml:74]: Building for multiple platforms is nice to have, as it allows easier usage on more servers. (Also lets people run it on a raspberry pi, etc.)
```suggestion
          platforms: linux/amd64,linux/arm64
          context: .
```

File: .github/workflows/docker-publish.yml
Comment (Zaprit) [inline .github/workflows/docker-publish.yml:94]: do we have to do any certificate setup for this, as we've not yet used cosign?

**Review by Zaprit — CHANGES_REQUESTED** [review]
Looks good for the most part, main thing is building ARM images.

File: .github/workflows/docker-publish.yml
Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:94]: I think this is something github does for you, though i could be wrong, ill double check

Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:94]: yeah github signs the containers for you

File: .github/workflows/docker-publish.yml
Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:74]: gimme a minute to mess around with this

Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:75]: @Zaprit & @sudokoko  it gives an exception during the "restore" phase, i don't know that much dotnet so im not sure exactly whats going on
Screenshot 2024-03-29 at 10 03 58 AM

Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:75]: logs: https://github.com/ren-makes-things/ProjectLighthouse/actions/runs/8483485830/job/23244693081

Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:75]: im going to rerun the job, see if it was just a fluke

Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:75]: fail's again, not sure whats happening, https://github.com/ren-makes-things/ProjectLighthouse/actions/runs/8483485830/job/23244818665

Comment (Zaprit) [reply .github/workflows/docker-publish.yml:75]: just looking at other pipelines that build for arm, seems like you need to do a setup qemu action, just before the buildx setup
```
      - name: Set up QEMU
        uses: docker/setup-qemu-action@v3
```

Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:75]: i'll give that a try and report back

Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:79]: https://github.com/ren-makes-things/ProjectLighthouse/actions/runs/8483690691/job/23245230894

Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:79]: @Zaprit any updates or anything i need to change on my end?

Comment (Zaprit) [reply .github/workflows/docker-publish.yml:79]: hmm, not sure at this point, seems to be a bug in .NET, there's no easy fix it seems, if you just remove the arm thing for now, I'll revisit it later.

Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:79]: i'll remove it for now

Comment (ren-makes-things) [reply .github/workflows/docker-publish.yml:74]: done

**Comment by ren-makes-things** [comment]
@Zaprit all done

**Review by Zaprit — APPROVED** [review]

**Pull request merged**
